### PR TITLE
Prepend `SyntaxStringInterpolationError` errors in macro expansion with `Internal macro error:`

### LINKED ISSUE #2039: `SyntaxStringInterpolationError` should prepend `Internal macro error: ` if hit by a macro expansion
### Description

If a macro expansion hits a `SyntaxStringInterpolationError`, we should prepend some appropriate text to the description to mark it as an internal macro error, eg. `Internal macro error: `.

### PR DESCRIPTION
**Pull request by natikgadzhi**
## Summary
This pull request prepends `SyntaxStringInterpolationError` message with `Internal macro error:`. Closes #2039.

## Changes
- In `MacroExpansionContext.addDiagnostics`, if the passed `error as? SyntaxStringInterpolationError` worked, we prepend it with the `Internal macro error` and grab it's description. Otherwise, we just use `String(describing:error)` as the diagnostic message.
- Adds a basic test for `addDiagnostics` for that behavior.

## Next steps
- @DougGregor, does this look like the right direction to you? I'm new to Macros, and I'm here to learn — no hard feelings if my attempt is naive / plain wrong.
- @Matejkob, sounds like your area of interest! Could you give me some hints to how I could test this end to end, and construct a test macro that would throw syntax string interpolation error?

### CODE REVIEW

File: Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift [hunk 1]
```diff
@@ -161,15 +161,18 @@ where Self.StringInterpolation == SyntaxStringInterpolation {
   init(stringInterpolation: SyntaxStringInterpolation)
 }
 
-enum SyntaxStringInterpolationError: Error, CustomStringConvertible {
+// Used in `MacroExpansionContext` to make a nicer error message when a
+// macro expansion fails with SwiftSyntaxBuilder making a syntax node
+// with string interpolation.
+@_spi(Diagnostics) public enum SyntaxStringInterpolationError: Error, CustomStringConvertible {
```
Comment (natikgadzhi) [inline Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:167]: Don't know if it's okay to `@_spi` this, but that seemed like a better option than to make it public outright.

File: Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift [hunk 2]
```diff
@@ -161,15 +161,18 @@ where Self.StringInterpolation == SyntaxStringInterpolation {
...
   case producedInvalidNodeType(expectedType: SyntaxProtocol.Type, actualType: SyntaxProtocol.Type)
   case diagnostics([Diagnostic], tree: Syntax)
 
   static func producedInvalidNodeType<S: SyntaxProtocol>(expectedType: SyntaxProtocol.Type, actualNode: S) -> Self {
     return .producedInvalidNodeType(expectedType: expectedType, actualType: type(of: actualNode))
   }
 
-  var description: String {
+  public var description: String {
```
Comment (natikgadzhi) [inline Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:175]: I _think_ compiler complained about this for custom string convertible.

File: Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+import SwiftSyntax
+import SwiftDiagnostics
+@_spi(Diagnostics) import SwiftSyntaxBuilder
+import SwiftSyntaxMacros
+import SwiftSyntaxMacroExpansion
+import SwiftSyntaxMacrosTestSupport
+import XCTest
+
+//private class TestMacroExpansionContext: MacroExpansionContext {
```
Comment (natikgadzhi) [inline Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift:29]: TODO: Remove the commented out test expansion context, basic context worked out fine.

File: Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:167]: I just checked and `SyntaxStringInterpolationError` currently serves two distinct use cases: 
1. The `.diagnostics` case in `SyntaxProtocol(validating:)`
2. The `.producedInvalidNodeType` case in SyntaxNodeWithBody and `DeclSyntaxProtocol.init`

I think what we should do, is split the error into two distinct structs and make the one that represents the `producedInvalidNodeType` case public. That way, we minimize the public surface are we are introducing while also allowing other clients to catch that case.

File: Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+final class MacroExpansionContextTests: XCTestCase {
+
+  func testMacroExpansionContextAddDiagnosticsAddsSwiftSyntaxInterpolationErrorsWithWrappingMessage(){
+    let context = BasicMacroExpansionContext()
+    let error = SyntaxStringInterpolationError.producedInvalidNodeType(expectedType: DeclSyntax.self, actualType: ExprSyntax.self)
+    // Since we only care about the error switch inside of addDagnostics, we don't care about the particular node we're passing in.
+    context.addDiagnostics(from: error, node: ExprSyntax("1"))
+
+    XCTAssertEqual(context.diagnostics.count, 1)
+    XCTAssertTrue(context.diagnostics.first!.message.starts(with: "Internal macro error:"))
```
Comment (ahoppen) [inline Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift:62]: I think you should be able to test the real-case scenario when this happens by defining an extension macro that returns 

```swift
try ExtensionDeclSyntax("var x: Int")
```

And then use `assertMacroExpansion` to check that the emitted error contains `Internal macro error:`

**Comment by natikgadzhi** [comment]
@ahoppen, thank you for the review! 

I've cleaned up the code and added an end-to-end test for now. The next part is splitting the error enum into two pieces. That'll be a separate commit.

**Comment by natikgadzhi** [comment]
@ahoppen ready for review!

File: Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift [hunk 5]
```diff
@@ -161,23 +161,28 @@ where Self.StringInterpolation == SyntaxStringInterpolation {
...
+  let diagnostics: [Diagnostic]
+  let tree: Syntax
 
   var description: String {
-    switch self {
-    case .producedInvalidNodeType(expectedType: let expectedType, actualType: let actualType):
-      return "Parsing the code snippet was expected to produce a \(expectedType) but produced a \(actualType)"
-    case .diagnostics(let diagnostics, let tree):
-      // Start the diagnostic on a new line so it isn't prefixed with the file, which messes up the
-      // column-aligned message from ``DiagnosticsFormatter``.
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:178]: Could we keep this comment?

File: Sources/SwiftSyntaxMacros/MacroExpansionContext.swift [hunk 6]
```diff
@@ -12,6 +12,7 @@
 
 import SwiftDiagnostics
 import SwiftSyntax
+@_spi(Diagnostics) import SwiftSyntaxBuilder
```
Comment (ahoppen) [inline Sources/SwiftSyntaxMacros/MacroExpansionContext.swift:15]: I don’t think we need `@_spi(Diagnostics)` anymore.

File: Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    providingExtensionsOf type: some TypeSyntaxProtocol,
+    conformingTo protocols: [TypeSyntax],
+    in context: some MacroExpansionContext
+  ) throws -> [ExtensionDeclSyntax] {
+    let ext = try ExtensionDeclSyntax("var x: Int")
+    return [ext]
+  }
+}
+
+final class MacroExpansionContextTests: XCTestCase {
```
Comment (ahoppen) [inline Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift:43]: I think a better name for this test would now be something like `StringInterpolationErrorTests`.

**Review by ahoppen — COMMENTED** [review]
A couple of minor comments, otherwise LGTM.

**Comment by natikgadzhi** [comment]
Oooops, those are embarrassing. I’ll clean it up.

File: Sources/SwiftSyntaxMacros/MacroExpansionContext.swift (hunk 6 above)
Comment (natikgadzhi) [reply Sources/SwiftSyntaxMacros/MacroExpansionContext.swift:15]: Good catch! Sorry I missed it.

**Comment by natikgadzhi** [comment]
@ahoppen done!

File: Tests/SwiftSyntaxMacroExpansionTest/StringInterpolationErrorTests.swift [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+final class StringInterpolationErrorTests: XCTestCase {
+
+  func testMacroExpansionContextAddDiagnosticsAddsSwiftSyntaxInterpolationErrorsWithWrappingMessage() {
+    let context = BasicMacroExpansionContext()
+    let error = SyntaxStringInterpolationInvalidNodeTypeError(expectedType: DeclSyntax.self, actualNode: ExprSyntax("test"))
+
+    // Since we only care about the error switch inside of addDagnostics, we don't care about the particular node we're passing in
+    context.addDiagnostics(from: error, node: ExprSyntax("1"))
+    XCTAssertEqual(context.diagnostics.count, 1)
+    XCTAssertTrue(context.diagnostics.first!.message.starts(with: "Internal macro error:"))
```
Comment (ahoppen) [inline Tests/SwiftSyntaxMacroExpansionTest/StringInterpolationErrorTests.swift:52]: Can you do the following? That way the test doesn’t crash if there is no diagnostic.

```suggestion
    let diagnostic = try XCTUnwrap(context.diagnostics.first)
    XCTAssertTrue(diagnostic.message.starts(with: "Internal macro error:"))
```

Same for the test below

**Review by ahoppen — COMMENTED** [review]
One minor comment, otherwise LGTM.

File: Tests/SwiftSyntaxMacroExpansionTest/StringInterpolationErrorTests.swift (hunk 8 above)
Comment (natikgadzhi) [reply Tests/SwiftSyntaxMacroExpansionTest/StringInterpolationErrorTests.swift]: Yep, good call. Fixed.

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by kimdv** [comment]
@swift-ci please test windows

**Comment by kimdv** [comment]
@swift-ci please test windows

**Comment by kimdv** [comment]
@swift-ci please test Windows

**Comment by kimdv** [comment]
@swift-ci please test Windows

**Pull request merged**